course3/homework-string/my_string.cpp

```diff
@@ -0,0 +1,209 @@
+        {
+            if(other == nullptr)
+            {
+                clearBuffer();
+            }
+            else
+            {
+                auto newSize =std::strlen(other) + 1;
+                auto newBuffer = std::make_unique<char[]>(newSize);
+
+                std::strcpy(newBuffer.get(), other);
+                mBuffer.swap(newBuffer);
+            }
+
+            return *this;
+        }
+
+        MyString& MyString::operator+=(const char* other)
+        {
+            if(other != nullptr && other[0] != '\0')
+            {
+                auto newSize = this->size() + std::strlen(other) + 1;
+                auto copyStartIndex = this->size();
+
+                auto newBuffer = std::make_unique<char[]>(newSize);
+                std::strcpy(newBuffer.get(), this->c_str());
+                std::strcpy(newBuffer.get() + copyStartIndex, other);
+
+                mBuffer.swap(newBuffer);
+            }
+
+            return *this;
+        }
+
+        MyString& MyString::operator+=(const MyString& other)
+        {
+            if(!other.empty())
+            {
+                auto newSize = this->size() + other.size() + 1;
+                auto copyStartIndex = this->size();
+
+                auto newBuffer = std::make_unique<char[]>(newSize);
+                std::strcpy(newBuffer.get(),this->c_str());
+                std::strcpy(newBuffer.get() + copyStartIndex, other.c_str());
+                mBuffer.swap(newBuffer);
+            }
+
+            return *this;
+        }
+
+        MyString MyString::operator+(const char* other) const
+        {
+            if(other == nullptr || other[0] == 0)
+            {
+                return MyString(*this);
+            }
+            else
+            {
+                auto newSize = this->size() + std::strlen(other) + 1;
+                auto copyStartIndex = this->size();
+                auto newBuffer = std::make_unique<char[]>(newSize);
+                std::strcpy(newBuffer.get(), mBuffer.get());
+                std::strcpy(newBuffer.get() + copyStartIndex, other);
+
+                auto result = MyString(newBuffer.get());
+                return result;
+            }
+        }
+
+        MyString MyString::operator+(const MyString& other) const
+        {
+            if(other.empty())
+            {
+                return MyString(*this);
+            }
+            else
+            {
+                auto newSize = this->size() + other.size() + 1;
+                auto copyStartIndex = this->size();
+
+                auto newBuffer = std::make_unique<char[]>(newSize);
+                std::strcpy(newBuffer.get(), mBuffer.get());
+                std::strcpy(newBuffer.get() + copyStartIndex, other.mBuffer.get());
+
+                return MyString(newBuffer.get());
+            }
+        }
+
+
+        void MyString::clearBuffer()
+        {
+            //TODO: what happens with the previous pointer?
```

Comment:
I meant the previous unique ptr. The comment is obsolete now - the previous value, if any, will go out of scope so it gets cleaned properly.